bitcoin/src/blockdata/witness.rs

```diff
@@ -166,26 +166,31 @@ impl Witness {
     }
 
     /// Creates [`Witness`] object from an array of byte-arrays
+    #[deprecated(since="0.30.0", note="use `witness::from_slice()` instead")]
     pub fn from_vec(vec: Vec<Vec<u8>>) -> Self {
-        let witness_elements = vec.len();
-        let index_size = witness_elements * 4;
+        Witness::from_slice(&vec)
+    }
 
-        let content_size: usize = vec
+    /// Creates a [`Witness`] object from a slice of bytes slices where each slice is a witness item.
+    pub fn from_slice<T: AsRef<[u8]>>(slice: &[T]) -> Self {
+        let witness_elements = slice.len();
+        let index_size = witness_elements * 4;
+        let content_size = slice
             .iter()
-            .map(|el| el.len() + VarInt(el.len() as u64).len())
+            .map(|elem| elem.as_ref().len() + VarInt(elem.as_ref().len() as u64).len())
             .sum();
+        
         let mut content = vec![0u8; content_size + index_size];
         let mut cursor = 0usize;
-        for (i, el) in vec.into_iter().enumerate() {
+        for (i, elem) in slice.iter().enumerate() {
             encode_cursor(&mut content, content_size, i, cursor);
-
-            let el_len_varint = VarInt(el.len() as u64);
-            el_len_varint
-                .consensus_encode(&mut &mut content[cursor..cursor + el_len_varint.len()])
+            let elem_len_varint = VarInt(elem.as_ref().len() as u64);
+            elem_len_varint
+                .consensus_encode(&mut &mut content[cursor..cursor + elem_len_varint.len()])
                 .expect("writers on vec don't errors, space granted by content_size");
-            cursor += el_len_varint.len();
-            content[cursor..cursor + el.len()].copy_from_slice(&el);
-            cursor += el.len();
+            cursor += elem_len_varint.len();
+            content[cursor..cursor + elem.as_ref().len()].copy_from_slice(elem.as_ref());
+            cursor += elem.as_ref().len();
```

Comment:
Out of curiosity, would cutting down on as_ref() calls increase performance at all?

Reply:
Not an expert but I don't expect it to have a big performance impact since `as_ref()` is just getting the reference.

Can anyone else comment?

Reply:
`as_ref` compiles down to essentially nothing for simple types and even if the impl is complicated, unless it internally uses `UnsafeCell` returning different pointers each time (:vomiting_face: extremely bad code) the compiler would very likely cache the result anyway.

However the change does improve performance because the instruction cache is not filled with redundant code - other code can fit in.